crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -104,10 +104,10 @@ pub fn derive_bundle(input: TokenStream) -> TokenStream {
     let mut field_from_components = Vec::new();
     for (field_type, field) in field_type.iter().zip(field.iter()) {
         field_component_ids.push(quote! {
-        <#field_type as #ecs_path::bundle::Bundle>::component_ids(components, storages, &mut *ids);
+        <#field_type as #ecs_path::bundle::Bundle>::component_ids(components, storages, &mut ids);
         });
         field_get_components.push(quote! {
-            self.#field.get_components(&mut *func);
+            self.#field.get_components(&mut func);
```

Comment:
And the reason for this API using &mut reveals itself.

The way you have implemented this, you are creating a tower of monomorphisation.
The root call is passed the closure, which is all fine.
But then that calls the next layer down with `&mut [the closure]`.
That in turns calls its next layer with `&mut &mut [the closure]`

The correct way to do what you want is to just pass `func`. Except, oh no, there's no way to generically reborrow func, so this doesn't work.